pkg/trace/sampler/remote_config_handler.go

```diff
@@ -0,0 +1,58 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package sampler
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state/products/apmsampling"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+)
+
+// ApmRemoteConfigHandler holds pointers to samplers that need to be updated when APM remote config changes
+type ApmRemoteConfigHandler struct {
+	remoteClient    config.RemoteClient
+	conf            *config.AgentConfig
+	prioritySampler *PrioritySampler
+}
+
+func NewApmRemoteConfigHandler(conf *config.AgentConfig, prioritySampler *PrioritySampler) *ApmRemoteConfigHandler {
+	if conf.RemoteSamplingClient == nil {
+		return nil
+	}
+
+	prioritySampler.EnableRemoteRates(conf.MaxRemoteTPS, conf.AgentVersion)
+
+	return &ApmRemoteConfigHandler{
+		remoteClient:    conf.RemoteSamplingClient,
+		conf:            conf,
+		prioritySampler: prioritySampler,
+	}
+}
+
+func (a *ApmRemoteConfigHandler) Start() {
+	if a == nil {
+		return
+	}
+
+	a.remoteClient.Start()
+	a.remoteClient.RegisterAPMUpdate(a.onUpdate)
+}
+
+func (a *ApmRemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	a.prioritySampler.remoteRates.update(update)
+	a.updateRareSamplerConfig(update)
+}
+
+func (a *ApmRemoteConfigHandler) updateRareSamplerConfig(update map[string]state.APMSamplingConfig) {
+	for _, conf := range update {
+		switch conf.Config.RareSamplerConfig {
+		case apmsampling.RareSamplerConfigEnabled:
+			a.conf.RareSamplerDisabled = false
```

Comment:
Do you know if the agent will automatically pick it up? Any concurrency issue we may have here? I am unfamiliar with the threading in place in the agent.

Reply:
I'm quite confident it will pick it up (here's the sampling code):

```
func (a *Agent) samplePriorityTrace(now time.Time, pt traceutil.ProcessedTrace) bool {
	var rare bool
	if a.conf.RareSamplerDisabled {
		rare = false
	} else {
		// run this early to make sure the signature gets counted by the RareSampler.
		rare = a.RareSampler.Sample(now, pt.TraceChunk, pt.TracerEnv)
	}
	if a.PrioritySampler.Sample(now, pt.TraceChunk, pt.Root, pt.TracerEnv, pt.ClientDroppedP0sWeight) {
		return true
	}
	if traceContainsError(pt.TraceChunk.Spans) {
		return a.ErrorsSampler.Sample(now, pt.TraceChunk.Spans, pt.Root, pt.TracerEnv)
	}
	return rare
}
```

This is the only usage of this config value so I see no concurrency issue, but I have nothing against being extra safe

Reply:
I see thanks for the clarification. I think it should be fine. You might have 2 threads, one writing and the other one reading. But I am not sure if it's an issue with bools.